src/internal/operators/groupBy.ts

```diff
@@ -127,214 +124,139 @@ export function groupBy<T, K, R>(
-  }
-
-  protected _next(value: T): void {
-    let key: K;
-    try {
-      key = this.keySelector(value);
-    } catch (err) {
-      this.error(err);
-      return;
-    }
-
-    this._group(value, key);
-  }
+            // Emit the grouped observable. Note that we can't do a simple `asObservable()` here,
+            // because the grouped observable has special semantics around reference counting
+            // to ensure we don't sever our connection to the source prematurely.
+            const grouped = createGroupedObservable(key, group);
+            subscriber.next(grouped);
 
-  private _group(value: T, key: K) {
-    let groups = this.groups;
+            if (durationSelector) {
+              const durationSubscriber = new OperatorSubscriber(
+                // Providing the group here ensures that it is disposed of -- via `unsubscribe` --
+                // wnen the duration subscription is torn down. That is important, because then
+                // if someone holds a handle to the grouped observable and tries to subscribe to it
+                // after the connection to the source has been severed, they will get an
+                // `ObjectUnsubscribedError` and know they can't possibly get any notifications.
+                group as any,
+                () => {
+                  // Our duration notified! We can complete the group.
+                  // The group will be removed from the map in the teardown phase.
+                  group!.complete();
+                  durationSubscriber?.unsubscribe();
+                },
+                undefined,
+                undefined,
+                // Teardown: Remove this group from our map.
+                () => groups.delete(key)
+              );
 
-    if (!groups) {
-      groups = this.groups = new Map<K, Subject<T | R>>();
-    }
+              // Start our duration notifier.
+              groupBySourceSubscriber.add(durationSelector(grouped).subscribe(durationSubscriber));
+            }
+          }
 
-    let group = groups.get(key);
+          // Send the value to our group.
+          group.next(elementSelector ? elementSelector(value) : value);
+        },
+        // Error from the source.
+        (err) => notify((consumer) => consumer.error(err)),
+        // Source completes.
+        () => notify((consumer) => consumer.complete()),
+        // Free up memory.
+        // When the source subscription is _finally_ torn down, release the subjects and keys
+        // in our groups Map, they may be quite large and we don't want to keep them around if we
+        // don't have to.
+        () => groups.clear()
+      );
 
-    let element: R;
-    if (this.elementSelector) {
-      try {
-        element = this.elementSelector(value);
-      } catch (err) {
-        this.error(err);
-      }
-    } else {
-      element = value as any;
-    }
+      // Subscribe to the source
+      return source.subscribe(groupBySourceSubscriber);
```

Comment:
issue: AFAICT, this `return` is not necessary. Won't it just try to add the `groupBySourceSubscriber` to the destination `subscriber`? IIRC, other operators don't do this - although, I'm inclined to check them, again, for unnecessary returns.

Reply:
That's correct. This isn't necessary.